pkg/kubelet/volumemanager/volume_manager_test.go

```diff
@@ -51,45 +54,103 @@ const (
 )
 
 func TestGetMountedVolumesForPodAndGetVolumesInUse(t *testing.T) {
-	tmpDir, err := utiltesting.MkTmpdir("volumeManagerTest")
-	if err != nil {
-		t.Fatalf("can't make a temp dir: %v", err)
-	}
-	defer os.RemoveAll(tmpDir)
-	cpm := podtest.NewMockCheckpointManager()
-	podManager := kubepod.NewBasicPodManager(podtest.NewFakeMirrorClient(), secret.NewFakeManager(), configmap.NewFakeManager(), cpm)
-
-	node, pod, pv, claim := createObjects()
-	kubeClient := fake.NewSimpleClientset(node, pod, pv, claim)
-
-	manager := newTestVolumeManager(tmpDir, podManager, kubeClient)
-
-	stopCh := runVolumeManager(manager)
-	defer close(stopCh)
-
-	podManager.SetPods([]*v1.Pod{pod})
-
-	// Fake node status update
-	go simulateVolumeInUseUpdate(
-		v1.UniqueVolumeName(node.Status.VolumesAttached[0].Name),
-		stopCh,
-		manager)
-
-	err = manager.WaitForAttachAndMount(pod)
-	if err != nil {
-		t.Errorf("Expected success: %v", err)
-	}
-
-	expectedMounted := pod.Spec.Volumes[0].Name
-	actualMounted := manager.GetMountedVolumesForPod(types.UniquePodName(pod.ObjectMeta.UID))
-	if _, ok := actualMounted[expectedMounted]; !ok || (len(actualMounted) != 1) {
-		t.Errorf("Expected %v to be mounted to pod but got %v", expectedMounted, actualMounted)
+	tests := []struct {
+		name                string
+		pvMode, podMode     v1.PersistentVolumeMode
+		disableBlockFeature bool
+		expectMount         bool
+		expectError         bool
+	}{
+		{
+			name:        "filesystem volume",
+			pvMode:      v1.PersistentVolumeFilesystem,
+			podMode:     v1.PersistentVolumeFilesystem,
+			expectMount: true,
+			expectError: false,
+		},
+		{
+			name:        "block volume",
+			pvMode:      v1.PersistentVolumeBlock,
+			podMode:     v1.PersistentVolumeBlock,
+			expectMount: true,
+			expectError: false,
+		},
+		{
+			name:                "block volume with block feature off",
+			pvMode:              v1.PersistentVolumeBlock,
+			podMode:             v1.PersistentVolumeBlock,
+			disableBlockFeature: true,
+			expectMount:         false,
+			expectError:         false,
+		},
+		{
+			name:        "mismatched volume",
+			pvMode:      v1.PersistentVolumeBlock,
+			podMode:     v1.PersistentVolumeFilesystem,
+			expectMount: false,
+			expectError: true,
```

Comment:
where does the error come from, is it a timeout or won't `expectedVolumes` have len 0?

Reply:
It's not timeout, DSWP has stored error into DSW and `WaitForAttachAndMount` asynchronously picked it up within ~`podAttachAndMountRetryInterval` (300ms)